### PR TITLE
feat(quinn-proto): Make compilable in wasm32-unknown-unknown and run wasm tests in CI

### PR DESCRIPTION
**Pull request by matheus23**
### Summary of Changes

- Use `web_time` dependency instead of `std::time` & re-export `Instant`, `Duration`, `SystemTime` and `UNIX_EPOCH` in `lib.rs` so imports are simpler
- Adjust imports to use `crate::{Instant, Duration, ...}` instead of `std::time::{Instant, Duration, ...}`
- Added `wasm-bindgen-test` to enable running tests from a Wasm build
- Added a CI task to test quinn-proto tests in Wasm with node.js

### Running `quinn-proto` tests in Wasm

```sh
# install wasm-bindgen-test-runner
$ cargo install wasm-bindgen-cli
# run tests
$ cargo test -p quinn-proto --target wasm32-unknown-unknown
```

<details>
<summary>Old open questions, now resolved</summary>

- Should the added code/dependencies additionally be feature-gated?
  
  I personally don't think so. As far as I know, there's practically no one asking for wasm32-unknown-unknown compilability who isn't targeting nodejs/browsers/cloudflare workers/etc. and in all of these cases wasm-bindgen is the de-facto standard.
  There's also targeting wasmtime and other runtimes, but IMO those needs are met by `wasm32-wasi-*` targets, not `wasm32-unknown-unknown`.
  
  That said, for my intents and purposes that's mostly a cosmetic question, I don't mind having to enable another feature flag. Other concerns could be "what non-web `wasm32-unknown-unknown` becomes a thing and we need to break compatibility by introducing features?".
- Should the `Duration`, `Instant`, etc. re-exports live in `lib.rs`?
  
  They're `pub(crate)` only, so I don't think it matters much. I can move them into a file if you want, or we keep them here.
- Should we just depend on `web_time` instead of `std::time` everywhere instead of cfg-gating & re-exporting?
  
  `web_time` itself already multiplexes between `std::time` and a web-based implementation. So in a way, we're doing it twice here. What we gain is not depending on `web_time` most of the time.
  However, once we work on Wasm support for e.g. `quinn` itself, we'd need to do this whole re-export dance, again. So perhaps it'd be better to just depend on `web_time` everywhere?
  
</details>

### CODE REVIEW

File: Cargo.toml
Comment (djc) [inline Cargo.toml:53]: Please keep the dependencies in alphabetic order. Also, it doesn't look like `rustls-pki-types` actually gets used here? And also please simplify `{ version = "x" }` to just `"x"`.

File: quinn-proto/Cargo.toml
Comment (djc) [inline quinn-proto/Cargo.toml:51]: Please add a comment here that we're only adding an edge for rustls-pki-types to activate the `web` feature.

File: .github/workflows/rust.yml
Comment (djc) [inline .github/workflows/rust.yml]: Pretty sure these crates names aren't quite right...

File: .github/workflows/rust.yml
Comment (djc) [inline .github/workflows/rust.yml:68]: I think we should just squash these changes into the first commit?

**Review by djc — COMMENTED** [review]
Please squash all the changes into a single commit, as these commits are not conceptually separate.

File: quinn-proto/src/tests/mod.rs
Comment (djc) [inline quinn-proto/src/tests/mod.rs:42]: I think we should find a way to avoid repeating these on every test.

**Review by djc — COMMENTED** [review]
This is looking pretty good!

File: quinn-proto/src/tests/mod.rs
Comment (matheus23) [reply quinn-proto/src/tests/mod.rs:42]: Unfortunately I don't know of a good way :/
What works:
- Writing our own proc macro
- Writing our own declarative macro (I think?)
- Splitting the test into 3 modules: one with `pub fn`s, one with `#[test]`s calling the `pub fn`s, and one with `#[wasm_bindgen_test]`s running the `pub fn`s.

I don't like either of these options. Would love some good ideas.

What doesn't work:
- Use only `#[wasm_bindgen_test]` (it doesn't resolve to just `#[test]` on other platforms, which is a shame IMO)

Comment (djc) [reply quinn-proto/src/tests/mod.rs:42]: > * Use only `#[wasm_bindgen_test]` (it doesn't resolve to just `#[test]` on other platforms, which is a shame IMO)

Have you investigated whether there is an open issue or some discussion of this in the upstream repo? It seems pretty straightforward...

Comment (matheus23) [reply quinn-proto/src/tests/mod.rs:42]: Yeah, I looked, there was nothing. Anyhow, [I opened an issue](https://github.com/rustwasm/wasm-bindgen/issues/4139).
Do you think we should block this PR on that issue & a subsequent release of wasm-bindgen-test?

Comment (djc) [reply quinn-proto/src/tests/mod.rs:42]: Let's see if there's any movement on your issue until Monday or something? Looks like at least one of the active maintainers in that repo is also a Quinn contributor (👋 @daxpedda), so let's see what happens.

Comment (daxpedda) [reply quinn-proto/src/tests/mod.rs:42]: Thank you for the ping djc!
https://github.com/rustwasm/wasm-bindgen/pull/4150, part of the latest release, should resolve this issue.
If not please let me know!

**Comment by matheus23** [comment]
Thanks for the review. Thoughts on this open question?

> Should we just depend on web_time instead of std::time everywhere instead of cfg-gating & re-exporting?

I'm on the fence. I think it may be better to depend on `web_time` on all targets. Otherwise I'll be repeating the same `lib.rs` changes in `quinn` itself (to avoid having `quinn_proto::Duration` be a public reexport).

**Comment by matheus23** [comment]
@djc @Ralith what's left to do here to get this over the finish line?
I'd love to also contribute `quinn` wasm32-unknown-unknown support after this.

**Review by djc — APPROVED** [review]
LGTM, please squash the commits?

**Comment by Ralith** [comment]
> I'm inclined to stick with depending on web_time only for WASM targets. IMO mainstream platforms should not bear the cost of niche platforms even if the overhead is small. @Ralith any thoughts?

Given that the amount of extra code is pretty trivial, this stance seems reasonable to me. I know I'd be a bit confused as a non-web user to see a web crate show up in my environment.

**Review by Ralith — APPROVED** [review]
The `wasm_bindgen_test` churn is unfortunate, but I guess there's nothing for it.

**Comment by djc** [comment]
@matheus23 this is currently failing the WASM test, I guess that's not supposed to be intentional?

**Comment by matheus23** [comment]
Seems like there's potentially some ring issues with wasm32.
Investigating this. Also need to make sure that `main` doesn't have any commits added that regress wasm32 support (So I need to merge in main continuously).

**Comment by matheus23** [comment]
Okay, got to somewhat of the bottom on this. The stacktraces you saw were v8 crashing for some reason.
Updating nodejs from 18 to 20 fixes the issue.

**Comment by matheus23** [comment]
Also seems like `cc` version 1.1.32 broke the ring Wasm build (but in github actions only?).
I've submitted an [upstream issue](https://github.com/rust-lang/cc-rs/issues/1275) and forced version 1.1.31 in CI for now.

File: quinn-proto/src/tests/mod.rs
Comment (matheus23) [reply quinn-proto/src/tests/mod.rs]: Not sure why I didn't think of this, but this works, too, at least in our case where we don't need async tests:
```rs
#[cfg(all(target_family = "wasm", target_os = "unknown"))]
use wasm_bindgen_test::wasm_bindgen_test as test;
```

I've changed the PR to use that. The diff is much smaller this way.

**Comment by matheus23** [comment]
FWIW, this PR is ready from my perspective. Is there anything left to do? Do you want to re-review?

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Thanks, this is looking good!

**Pull request merged**
